src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,337 @@
+        /// <param name="nodeStats">Registers an action used to append node stats when collected.</param>
+        /// <param name="asyncLoopFactory">Factory for creating and also possibly starting application defined tasks inside async loop.</param>
+        public ProvenBlockHeaderStore(
+            ConcurrentChain chain,
+            IDateTimeProvider dateTimeProvider,
+            ILoggerFactory loggerFactory,
+            IProvenBlockHeaderRepository provenBlockHeaderRepository,
+            INodeLifetime nodeLifetime,
+            INodeStats nodeStats,
+            IAsyncLoopFactory asyncLoopFactory)
+        {
+            Guard.NotNull(chain, nameof(chain));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            Guard.NotNull(provenBlockHeaderRepository, nameof(provenBlockHeaderRepository));
+            Guard.NotNull(nodeLifetime, nameof(nodeLifetime));
+            Guard.NotNull(nodeStats, nameof(nodeStats));
+            Guard.NotNull(asyncLoopFactory, nameof(asyncLoopFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.chain = chain;
+            this.provenBlockHeaderRepository = provenBlockHeaderRepository;
+            this.nodeLifetime = nodeLifetime;
+
+            this.lockObject = new object();
+            this.PendingBatch = new Dictionary<int, ProvenBlockHeader>();
+            this.Cache = new MemorySizeCache<int, ProvenBlockHeader>(this.MemoryCacheSizeLimitInBytes);
+
+            this.asyncLoopFactory = asyncLoopFactory;
+
+            this.performanceCounter = new BackendPerformanceCounter(dateTimeProvider);
+            nodeStats.RegisterStats(this.AddBenchStats, StatsType.Benchmark, 100);
+        }
+
+        /// <inheritdoc />
+        public async Task InitializeAsync()
+        {
+            await this.provenBlockHeaderRepository.InitializeAsync().ConfigureAwait(false);
+
+            this.TipHashHeight = this.provenBlockHeaderRepository.TipHashHeight;
+
+            this.logger.LogDebug("Initialized ProvenBlockHeader block tip at '{0}'.", this.TipHashHeight);
+
+            this.asyncLoop = this.asyncLoopFactory.Run("ProvenBlockHeaders job", async token =>
+            {
+                // Save pending items.
+                await this.SaveAsync().ConfigureAwait(false);
+
+                this.logger.LogTrace("(-)[IN_ASYNC_LOOP]");
+            },
+            this.nodeLifetime.ApplicationStopping,
+            repeatEvery: TimeSpan.FromMinutes(1),
+            startAfter: TimeSpan.FromMinutes(1));
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            ProvenBlockHeader header = null;
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                this.Cache.TryGetValue(blockHeight, out header);
+
+                if (header == null)
```

Comment:
check bool returned by TryGetValue, not out param against null

Reply:
Ok.